### ISSUE TITLE
afl-analyze not working!

### ISSUE THREAD

**Issue by pinkhat-m** [aflnet/aflnet#50]
Hello,
I am running a test with AFLNet. AFLNet could discover several crashes in our example, and I would like to use afl-analyze to get some analysis information for the crashes or other inputs, however, I am receiving an error like this:
```
[-] PROGRAM ABORT : No instrumentation detected.
         Location : main(), afl-analyze.c:1077
```
We instrument our code in order to perform the AFLNet test. Please clarify why I am unable to utilize this tool.

Thank you in advance.

**Comment by thuanpv** [comment]
Hi,
It is nice to know that AFLNet has found several crashes in your example.

We have not made any changes to afl-analyze yet so it is reasonable that it cannot be used as-is to analyze test cases generated by AFLNet. To make it work, the following major changes would need to be done:
1. Support AFLNet arguments (e.g., for network settings)
2. Add the send_over_network function to afl-analyze & update the logic accordingly so that can communicate with the server under test over network.

We can make the changes later but for now, you can use aflnet-replay to reproduce and analyze the results.

Regards,
Thuan

**Comment by pinkhat-m** [comment]
Hello,

Sure, thank you for your answer.

Respectfully yours,
Mahdis.

**Comment by pinkhat-m** [comment]
I have another question regarding parallel fuzzing.

I was attempting to run two AFLNet tests at the same time, but I was unable to do so since fuzzer failed with a "segmentation fault" message!

Should AFLNet be able to perform parallel fuzzing?

Respectfully yours,
Mahdis.

**Comment by thuanpv** [comment]
We have not tested the parallel fuzzing mode yet but technically it would work. I think the reason why you got that error message is because all the fuzzing instances try to connect to the same server at the same port. If this is the case, you might want to run different servers on different ports and update fuzzing commands accordingly. It becomes more complicated if the server uses some shared folder, like in case of FTP servers. For those, you might also need to update the configuration file of the server -- each server instance has its own configuration file in which the shared folder is set.

Another approach could be that you leverage our [ProFuzzBench](https://github.com/profuzzbench/profuzzbench) platform to run fuzzing instances on different containers and share the findings with each other.

I hope it helps.

**Comment by pinkhat-m** [comment]
Thank you for your response, I will try it.

**Comment by pinkhat-m** [comment]
I attempted to use afl-tmin on one of the input files, but received the following error:
```
[-] PROGRAM ABORT : No instrumentation detected.
         Location : main(), afl-tmin.c:1141

```
Is it for the same reason that afl-analyze is?

**Comment by pinkhat-m** [comment]
oh! I was running the wrong command for afl-analyze! I should inform you that it is working very well, I suppose!! (at least I can see some output!)

**Comment by pinkhat-m** [comment]
Hello,
I have a question regarding afl-tmin and afl-cmin. Have you implemented them for AFLNet fuzzer? (I have seen you didn't change their code since 2 years ago)
The results that I am getting do not make sense at all.
For example, for the afl-cmin it minimizes the queue from 1377 to just 1 test case, and I am receiving this result when the fuzzer still could find new crashes! I am not sure if it checks anything like coverage or trace in our application.

For afl-tmin is somehow the same result. below is our result which changes the size to zero!
```
[+] Read 1452230 bytes from 'id:001314,src:000311+000936,op:splice,rep:32'.
[*] Performing dry run (mem limit = 500 MB, timeout = 1000 ms)...
[+] Program terminates normally, minimizing in instrumented mode.
[*] Stage #0: One-time block normalization...
[+] Block normalization complete, 1452230 bytes replaced.
[*] --- Pass #1 ---
[*] Stage #1: Removing blocks of data...
    Block length = 131072, remaining size = 1452230
[+] Block removal complete, 1452230 bytes deleted.
[!] WARNING: Down to zero bytes - check the command line and mem limit!
[*] Stage #2: Minimizing symbols (0 code points)...
[+] Symbol minimization finished, 0 symbols (0 bytes) replaced.
[*] Stage #3: Character minimization...
[+] Character minimization done, 0 bytes replaced.
[*] --- Pass #2 ---
[*] Stage #1: Removing blocks of data...
    Block length = 1, remaining size = 0
[+] Block removal complete, 0 bytes deleted.

     File size reduced by : 100.00% (to 0 bytes)
    Characters simplified : 145223000.00%
     Number of execs done : 102
          Fruitless execs : path=0 crash=0 hang=0

[*] Writing output to 'minimized_input'...
[+] We're done here. Have a nice day!
```
I even tried to increase the memory but didn't affect it.
Could you please let me know if they are not applied on AFLNet?

Thank you.

**Comment by pinkhat-m** [comment]
These findings have me concerned about the afl-analyze as well. I am not sure if I really understood it, but I saw some similarities when comparing it to the Wireshark output, though I am not sure whether this is the expected outcome. (This file hasn't modified in over two years)

**Comment by thuanpv** [comment]
Hi @pinkhat-m,

I think I have answered your question in this thread. See my first response.

We have not made any changes to afl-analyze yet so it is reasonable that it cannot be used as-is to analyze test cases generated by AFLNet in some meaningful way. To make it work, the following main changes would need to be done:

- Support AFLNet's afl-fuzz arguments (e.g., for network settings)
- Add the send_over_network function to afl-analyze.c & update the logic accordingly so that it can communicate with the server under test over network.

**Comment by pinkhat-m** [comment]
Yes, but when I could see some result I thought that it was just because I was running it wrongly. And also I wanted to be sure that this is the same situation for afl-tmin and afl-cmin, which it is.

Thank you for the answer.

Another thing is that in "pend fav" on the status screen, sometimes I have seen the number like "4.29G", could you please tell me what it could be? (because it does not make sense to me)

And I would appreciate it if you could give me more explanation on the second number in "now processing" and also "total tmouts"; at the beginning, I thought that they are related to each other, based on the AFL explanation, but apparently they are not.

**Comment by pinkhat-m** [comment]
One test is now running; it displays "havoc" in "now trying" but there is no change in the "havoc" section! Is there a difference between these two?

**Comment by pinkhat-m** [comment]
From one test, I have reproduced the crashes, it is strange that I could see "Transfer-Encoding: chunked" in the client request which is related to HTTP header, while there is no HTTP dictionary in the AFL or AFLNet, I suppose. Could you please tell me how is that possible?

**Comment by pinkhat-m** [comment]
Hello, I have another question.
In one test I did not use -R option, but still I could see the "region" folder in the output directory; is it going to be created always but not use in the mutation??

**Comment by pinkhat-m** [comment]
oh still there is another one!
-K is supposed to be optional as you mentioned in the repository but I cannot run a test without it. I think it should be mandatory to use it yes?

**Comment by pinkhat-m** [comment]
How can I find some explanation about different selection strategies except inside the code?

**Comment by pinkhat-m** [comment]
I would like to know more about the name of crash file, if it is possible.
``` 
\:000000\ , sig\:11\ ,src \:000048+000882\ , op\: splice\,rep\:4 
``` 
\:000000      this could be the file number
op\: splice\      this is the mutation operation

I know when we have havoc in op, we have one file in src; which I think havoc is applied on src test case and then generate new one.
but in splice, that is going to take two different inputs from two locations inside the queue and put them in one location, I could imagine that there is no mutation operation on the new test case, but it should come from somewhere yes? could you please tell me how new test case is generated?

- Then I would like to know about sig and rep(as I see it is just a power of two).

**Comment by pinkhat-m** [comment]
Hello,
I suppose "max_depth" in plot_data file is the "levels". I cannot see why in AFLNet it always starts from 2. Could you please explain why?
